flink-ml-lib/src/main/java/org/apache/flink/ml/anomalydetection/isolationforest/IsolationForest.java

```diff
@@ -0,0 +1,611 @@
+                }
+            }
+
+            return pathLength + calculateCn(tmpITree.leafNodesNum);
+        }
+
+        private double calculateCn(double n) {
+            if (n <= 1) {
+                return 0;
+            }
+            return 2.0 * (Math.log(n - 1.0) + 0.5772156649015329) - 2.0 * (n - 1.0) / n;
+        }
+
+        private DenseVector classifyByCluster(DenseVector scores) {
+            int scoresSize = scores.size();
+            this.center0 = scores.get(0); // Cluster center of abnormal
+            this.center1 = scores.get(0); // Cluster center of normal
+
+            for (int p = 1; p < scores.size(); p++) {
+                if (scores.get(p) > center0) {
+                    center0 = scores.get(p);
+                }
+
+                if (scores.get(p) < center1) {
+                    center1 = scores.get(p);
+                }
+            }
+
+            int cnt0, cnt1;
+            double diff0, diff1;
+            int[] labels = new int[scoresSize];
+
+            for (int i = 0; i < iters; i++) {
+                cnt0 = 0;
+                cnt1 = 0;
+
+                for (int j = 0; j < scoresSize; j++) {
+                    diff0 = Math.abs(scores.get(j) - center0);
+                    diff1 = Math.abs(scores.get(j) - center1);
+
+                    if (diff0 < diff1) {
+                        labels[j] = 0;
+                        cnt0++;
+                    } else {
+                        labels[j] = 1;
+                        cnt1++;
+                    }
+                }
+
+                diff0 = center0;
+                diff1 = center1;
+
+                center0 = 0.0;
+                center1 = 0.0;
+                for (int k = 0; k < scoresSize; k++) {
+                    if (labels[k] == 0) {
+                        center0 += scores.get(k);
+                    } else {
+                        center1 += scores.get(k);
+                    }
+                }
+
+                center0 /= cnt0;
+                center1 /= cnt1;
+
+                if (center0 - diff0 <= 1e-6 && center1 - diff1 <= 1e-6) {
+                    break;
+                }
+            }
+            return new DenseVector(new double[] {center0, center1});
+        }
+    }
+
+    /** Construct isolation tree. */
+    public static class ITree implements Serializable {
+        public int attributeIndex;
+        public double splitAttributeValue;
+        public ITree leftTree, rightTree;
+        public int currentHeight;
+        public int leafNodesNum;
+
+        public ITree() {}
+
+        public ITree(int attributeIndex, double splitAttributeValue) {
+            this.attributeIndex = attributeIndex;
+            this.splitAttributeValue = splitAttributeValue;
+            this.leftTree = null;
+            this.rightTree = null;
+            this.currentHeight = 0;
+            this.leafNodesNum = 1;
+        }
+
+        public static ITree createITree(
```

Comment:
If `ITree` is used to describe model data, it might be better to move the logic of computing/processing `ITree` out of this class.

Ideally, the description of the model data should be put in the `IsolationForestModelData.java` and the training logic should be put in the corresponding estimator class.

Same for `IForest`.

Reply:
OK, thanks for your advice and review teacher, I will fix them one by one.